### PR TITLE
[#1202] Domainmodel example: Add duplicated property validation rule.

### PR DESCRIPTION
**Pull request by miklossy**
- Add checkPropertyNamesAreUnique check to the DomainmodelValidator
class.
- Add DUPLICATE_PROPETY to the IssueCodes.
- Implement corresponding ValidationTests test cases.

Signed-off-by: Tamas Miklossy <miklossy@itemis.de>

### CODE REVIEW

**Comment by kthoms** [comment]
Would it be possible to use `NamesAreUniqueValidator` instead?

**Comment by cdietrich** [comment]
@kthoms didnt we remove this one lately explicitely?

**Comment by miklossy** [comment]
Yes, we removed that one from the Web Example because the `NamesAreUniqueValidator` does not work in combination with Xbase. See [commit](https://github.com/eclipse/xtext-web/commit/86ba27af23f71a9aabfd0f0ec4f000152a8bbb94) from @cdietrich

**Comment by LorenzoBettini** [comment]
IIRC `NamesAreUniqueValidator` doesn't work with Xbase because it would detect an inferred class name and an inferred constructor as duplicates (because they have the same name), since it acts directly on the inferred Jvm model, am I right?

Concerning this PR, I think it would be nice to have a similar duplicate check also for `Operation`. In that case, it would be nice to use the same Java mechanisms for NOT considering as duplicates operations with a different overloaded signature. With that respect, one could borrow from what Xtend does, by acting on the inferred Jvm model and by performing checks also taking into consideration type erasure. Xbase provides `org.eclipse.xtext.xbase.typesystem.^override.OverrideHelper` for that. I've done that in the past (https://github.com/LorenzoBettini/javamm/blob/master/javamm/src/javamm/util/JavammModelUtil.xtend).

Please let me know if you think it's worthwhile to do that in this example, possibly in a separate PR I can (hopefully) work on.

**Comment by miklossy** [comment]
@LorenzoBettini I think it is a good idea. I created a [separate issue](https://github.com/eclipse/xtext-eclipse/issues/1205) for that and assigned to you.
Would be awesome if you could provide a PR for that.

**Comment by LorenzoBettini** [comment]
@miklossy OK, I'll try to do that ASAP! I haven't updated my development workspace for some time now, I hope that Oomph will be kind... ;)

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel/src/org/eclipse/xtext/example/domainmodel/validation/DomainmodelValidator.xtend [hunk 1]
```diff
@@ -50,4 +53,20 @@ class DomainmodelValidator extends AbstractDomainmodelValidator {
 		}
 	}
 
+	@Check def void checkPropertyNamesAreUnique(Entity entity) {
+		val Multimap<String, Property> name2properties = HashMultimap.create
+
+		entity.features.filter(Property).filter[name!==null && !name.isEmpty].forEach[
```
Comment (kthoms) [inline org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel/src/org/eclipse/xtext/example/domainmodel/validation/DomainmodelValidator.xtend]: name.nullOrEmpty

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel/src/org/eclipse/xtext/example/domainmodel/validation/DomainmodelValidator.xtend [hunk 2]
```diff
@@ -50,4 +53,20 @@ class DomainmodelValidator extends AbstractDomainmodelValidator {
...
+			name2properties.put(name, it)
+		]
+
+		name2properties.keySet.forEach[
```
Comment (kthoms) [inline org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel/src/org/eclipse/xtext/example/domainmodel/validation/DomainmodelValidator.xtend]: you actually don't need to iterate over keySet and then use the key to get an entry, you coud instead directly iterate over the values

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel/src/org/eclipse/xtext/example/domainmodel/validation/IssueCodes.java [hunk 3]
```diff
@@ -20,4 +20,6 @@
 
 	String MISSING_TYPE = PREFIX + "MissingType";
 
+	String DUPLICATE_PROPETY = PREFIX + "duplicate_property";
```
Comment (kthoms) [inline org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel/src/org/eclipse/xtext/example/domainmodel/validation/IssueCodes.java]: Why did you decide to use different casing than above?

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel/src/org/eclipse/xtext/example/domainmodel/validation/DomainmodelValidator.xtend (hunk 1 above)
Comment (miklossy) [reply org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel/src/org/eclipse/xtext/example/domainmodel/validation/DomainmodelValidator.xtend]: Fixed.

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel/src/org/eclipse/xtext/example/domainmodel/validation/IssueCodes.java (hunk 3 above)
Comment (miklossy) [reply org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel/src/org/eclipse/xtext/example/domainmodel/validation/IssueCodes.java]: Fixed.

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel/src/org/eclipse/xtext/example/domainmodel/validation/DomainmodelValidator.xtend (hunk 2 above)
Comment (miklossy) [reply org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel/src/org/eclipse/xtext/example/domainmodel/validation/DomainmodelValidator.xtend]: Fixed.

**Pull request merged**